### DISCUSSION TITLE
Separating "draft" and "live" models

### DISCUSSION THREAD

**Discussion by sldblog** [ministryofjustice/hmpps-interventions-service#893] [Ideas]
The `referral_detail` refactoring (#891) highlighted that drafts behave differently from "sent" or "live" referrals.

## Observations:

- We never use "draft referrals" outside its own process.
- We have to choose if the new `referral_detail` table stores a single, updateable version during draft phase or stores separate versions for each draft UI page. (Draft and sent behaviours are potentially different.)
- Our API endpoints specifically separate drafts from sent referrals. (Which also means we have specific code paths.)

## Conclusion:

**We should separate draft models from sent/live models.**

## Benefits of separation:

- Clearer data models.
- Simpler queries (no more `sent_at IS NOT NULL` everywhere).
- The question of how `referral_details` behaves during drafts and after drafts no longer exists.

## Disadvantages of separation:

- Data duplication.

Thoughts?

**Comment by tomsmyers** [comment]
i think we can aleviate the issue of data duplication by deleting the drafts once the referral is successfully sent. otherwise i think this will greatly simplify not only the model, but also the API as we move to allowing sent referrals to be updated.

**Reply by sldblog** [reply]
Yeah. 💭 We will probably have to respond with a "410 Gone" when people try to update an already sent draft, to make it obvious it didn't just "disappear"

**Comment by sldblog** [comment]
cc @kth13 @OmotayoTalabi

**Comment by kth13** [comment]
Splitting Live and Draft Referrals initial implementation details
  * Draft referrals table will acts as primary table holding the foreign referential integrity for the following
      * selectedServiceCategories,
      * selectedDesiredOutcomes,
      * complexityLevelIds,
      * draft_oasys_risk_information,
      * referral_details
* The existing referral is copied into draft referral
* The existing draft referrals present in the referral is deleted once the migration is completed
* The draft and live referral holds the same id
* When the draft referral is moved to the live referral (referral table), the draft referral will not be deleted at the moment
